### PR TITLE
move to multi process tests

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

File: tests/flow/test_stress.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import os
 import sys
 import time
-import threading
+from pathos.pools import ProcessPool as Pool
```
Comment (swilly22) [inline tests/flow/test_stress.py:4]: order lines by length

File: tests/flow/test_stress.py [hunk 2]
```diff
@@ -71,6 +72,10 @@ def BGSAVE_loop(env, conn, n_iterations):
 
 class testStressFlow(FlowTestsBase):
     def __init__(self):
+        # skip test if we're running under Valgrind
+        if Env().envRunner.debugger is not None:
+            Env().skip() # valgrind is not working correctly with replication
```
Comment (swilly22) [inline tests/flow/test_stress.py:77]: not sure if this comment is relevant

File: tests/flow/test_stress.py [hunk 3]
```diff
@@ -23,8 +23,9 @@ def query_crud(graph, threadID):
             graph.query(read_query)
             graph.query(update_query)
             graph.query(delete_query)
+            return True
```
Comment (swilly22) [inline tests/flow/test_stress.py:26]: `def query_crud(graph, threadID):`
please rename` threadID` to something more relevant/meaningful

File: tests/flow/test_stress.py [hunk 4]
```diff
@@ -85,18 +90,14 @@ def __del__(self):
...
-            t = threading.Thread(target=query_crud, args=(graph, i))
-            t.setDaemon(True)
-            threads.append(t)
-            t.start()
-
-        # wait for threads to return
-        for i in range(CLIENT_COUNT):
-            t = threads[i]
-            t.join()
+        indexes = range(CLIENT_COUNT)
```
Comment (swilly22) [inline tests/flow/test_stress.py:93]: Set `CLIENT_COUNT` to some multiple of `GRAPH.CONFIG GET THREAD_COUNT`

File: tests/flow/test_stress.py [hunk 5]
```diff
@@ -85,18 +90,14 @@ def __del__(self):
...
+        pool = Pool(nodes=CLIENT_COUNT)
+
+        # invoke queries
+        m = pool.amap(query_crud, graphs, indexes)
+
+        # wait for processes to return
+        m.wait()
```
Comment (swilly22) [inline tests/flow/test_stress.py:100]: switch to `pool.map` which should be synchronized, in which case no need to call `m.wait`

File: tests/flow/test_stress.py [hunk 6]
```diff
@@ -113,59 +114,45 @@ def test01_bgsave_stress(self):
         conn = self.env.getConnection()
         graphs[0].query("CREATE INDEX ON :Node(val)")
```
Comment (swilly22) [inline tests/flow/test_stress.py:115]: line 110, same request regarding the comment

File: tests/flow/test_stress.py [hunk 7]
```diff
@@ -113,59 +114,45 @@ def test01_bgsave_stress(self):
...
-        t2.join()
-        t3.join()
-        t4.join()
-        t5.join()
+        # wait for processes to join
+        t1.wait()
+        t2.wait()
+        t3.wait()
+        t4.wait()
+        t5.wait()
```
Comment (swilly22) [inline tests/flow/test_stress.py:134]: is there a function within `pool` that would wait for all of the outstanding async operations ?

File: tests/flow/test_stress.py [hunk 8]
```diff
@@ -113,59 +114,45 @@ def test01_bgsave_stress(self):
...
+    #     m = pool.amap(query_crud, graphs, indexes)
+
+    #     # sleep for half a second, allowing threads to kick in
+    #     sleep(0.2)
+
+    #     conn = self.env.getConnection()
+    #     conn.shutdown()
+
+    #     # wait for processes to return
+    #     m.wait()
```
Comment (swilly22) [inline tests/flow/test_stress.py:155]: I think we should still run this code although we can't check exit code.

File: tests/flow/test_query_mem_limit.py [hunk 9]
```diff
@@ -1,6 +1,5 @@
 import random
-import threading
-from itertools import cycle
+from pathos.pools import ProcessPool as Pool
 from RLTest import Env
 from redisgraph import Graph
```
Comment (swilly22) [inline tests/flow/test_query_mem_limit.py:4]: order lines by length

File: tests/flow/test_query_mem_limit.py [hunk 10]
```diff
@@ -30,45 +29,50 @@
...
+    except Exception as e:
+        assert "Query's mem consumption exceeded capacity" in str(e)
+        return should_fail
+
 class testQueryMemoryLimit():
     def __init__(self):
-        global g
+        # skip test if we're running under Valgrind
+        if Env().envRunner.debugger is not None:
+            Env().skip() # valgrind is not working correctly with replication
```
Comment (swilly22) [inline tests/flow/test_query_mem_limit.py:45]: not sure if this comment is relevant

File: tests/flow/test_query_mem_limit.py [hunk 11]
```diff
@@ -30,45 +29,50 @@
...
-            t = threading.Thread(target=self.issue_query, args=(con, q[0], q[1]))
-            t.setDaemon(True)
-            threads.append(t)
-            t.start()
-
-        # wait for threads to return
-        while len(threads) > 0:
-            t = threads.pop()
-            t.join()
+        m = pool.amap(issue_query, connections, qs, should_fails)
```
Comment (swilly22) [inline tests/flow/test_query_mem_limit.py:66]: Are we sure `len(connections) == len(qs)` ? if so can we make the construction more obvious ?

File: tests/flow/test_query_mem_limit.py [hunk 12]
```diff
@@ -30,45 +29,50 @@
...
+
+        # wait for processes to return
+        m.wait()
```
Comment (swilly22) [inline tests/flow/test_query_mem_limit.py:69]: consider switching to `pool.map` and removing this line

File: tests/flow/test_pending_queries_limit.py [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import threading
+from pathos.pools import ProcessPool as Pool
 from RLTest import Env
 from redisgraph import Graph
```
Comment (swilly22) [inline tests/flow/test_pending_queries_limit.py:3]: oder lines by length

File: tests/flow/test_pending_queries_limit.py [hunk 14]
```diff
@@ -8,22 +8,23 @@
...
     except Exception as e:
         assert "Max pending queries exceeded" in str(e)
-        error_encountered = True
+        return True
 
 class testPendingQueryLimit():
     def __init__(self):
+        # skip test if we're running under Valgrind
+        if Env().envRunner.debugger is not None:
+            Env().skip() # valgrind is not working correctly with replication
```
Comment (swilly22) [inline tests/flow/test_pending_queries_limit.py:26]: not sure if this comment is relevant

File: tests/flow/test_pending_queries_limit.py [hunk 15]
```diff
@@ -42,48 +43,43 @@ def test_01_query_limit_config(self):
         self.env.assertEquals(max_queued_queries, 10)
 
     def stress_server(self):
-        threads = []
         connections = []
+        qs = []
```
Comment (swilly22) [inline tests/flow/test_pending_queries_limit.py:47]: swap lines

File: tests/flow/test_pending_queries_limit.py [hunk 16]
```diff
@@ -42,48 +43,43 @@ def test_01_query_limit_config(self):
...
         threadpool_size = self.conn.execute_command("GRAPH.CONFIG", "GET", "THREAD_COUNT")[1]
-        thread_count = threadpool_size * 5
+        thread_count = threadpool_size * 2
+        pool = Pool(nodes=thread_count)
 
         # init connections
         for i in range(thread_count):
             connections.append(self.env.getConnection())
+            qs.append(SLOW_QUERY)
```
Comment (swilly22) [inline tests/flow/test_pending_queries_limit.py:55]:
```python
qa = [SLOW_QUERY] * thread_count
connections = [self.env.getConnection()] * thread_count
```

File: tests/flow/test_pending_queries_limit.py [hunk 17]
```diff
@@ -42,48 +43,43 @@ def test_01_query_limit_config(self):
...
-            threads.append(t)
-            t.start()
+        m = pool.amap(issue_query, connections, qs)
 
-        # wait for threads to return
-        for i in range(thread_count):
-            t = threads[i]
-            t.join()
+        # wait for processes to return
+        m.wait()
```
Comment (swilly22) [inline tests/flow/test_pending_queries_limit.py:61]: consider switching to `pool.map` and removing this call

File: tests/flow/test_concurrent_query.py [hunk 18]
```diff
@@ -1,6 +1,4 @@
-import os
-import sys
-import threading
+from pathos.pools import ProcessPool as Pool
 from RLTest import Env
 from redisgraph import Graph, Node, Edge
 from redis import ResponseError
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:4]: order lines by length

File: tests/flow/test_concurrent_query.py [hunk 19]
```diff
@@ -10,64 +8,74 @@
...
         # Graph deletion failed.
-        assertions[threadID] = False
+        return False
+
+def run_concurrent(env, queries, f):
+    indexes = range(CLIENT_COUNT)
+    pool = Pool(nodes=CLIENT_COUNT)
+
+    # invoke queries
+    m = pool.amap(f, graphs, queries, indexes)
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:61]: it is unclear that `len(indexes) == len(queries)`
consider: `indexes = range(len(queries))`

File: tests/flow/test_concurrent_query.py [hunk 20]
```diff
@@ -10,64 +8,74 @@
...
+
+    # wait for processes to return
+    m.wait()
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:64]: consider switching to `pool.map`

File: tests/flow/test_concurrent_query.py [hunk 21]
```diff
@@ -10,64 +8,74 @@
...
+    result = m.get()
+
+    # validate all process return true
+    env.assertTrue(all(result))
 
 class testConcurrentQueryFlow(FlowTestsBase):
     def __init__(self):
+        # skip test if we're running under Valgrind
+        if Env().envRunner.debugger is not None:
+            Env().skip() # valgrind is not working correctly with replication
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:77]: not sure if this comment is relevant

File: tests/flow/test_concurrent_query.py [hunk 22]
```diff
@@ -98,68 +106,33 @@ def populate_graph(self):
...
-            t = threads[i]
-            t.join()
+        indexes = range(CLIENT_COUNT)
+        pool = Pool(nodes=CLIENT_COUNT)
+
+        # invoke queries
+        m = pool.amap(delete_graph, graphs, indexes)
+
+        # wait for processes to return
+        m.wait()
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:132]: switch to `pool.map`

File: tests/flow/test_concurrent_query.py [hunk 23]
```diff
@@ -10,64 +8,74 @@
...
 def thread_run_query(graph, query, threadID):
-    global assertions
     try:
-        assertions[threadID] = graph.query(query)
+        return graph.query(query)
     except ResponseError as e:
-        exceptions[threadID] = str(e)
+        return str(e)
 
 def delete_graph(graph, threadID):
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:47]: no need for threadID

File: tests/flow/test_concurrent_query.py [hunk 24]
```diff
@@ -98,68 +106,33 @@ def populate_graph(self):
...
-        
-        # Wait for threads to return.
-        for i in range(CLIENT_COUNT):
-            t = threads[i]
-            t.join()
+        indexes = range(CLIENT_COUNT)
+        pool = Pool(nodes=CLIENT_COUNT)
+
+        # invoke queries
+        m = pool.amap(delete_graph, graphs, indexes)
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:129]: no need for `indexes`

File: tests/flow/test_concurrent_query.py [hunk 25]
```diff
@@ -10,64 +8,74 @@
...
     for i in range(10):
         actual_result = graph.query(query)        
         if actual_result.nodes_created != 1 or actual_result.properties_set != 1:
-            assertions[threadID] = False
-            break
+            return False
+
+    return True
 
 def thread_run_query(graph, query, threadID):
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:41]: no need for `threadID`

File: tests/flow/test_concurrent_query.py [hunk 26]
```diff
@@ -172,22 +145,25 @@ def test_05_concurrent_read_delete(self):
...
-            self.env.assertEquals(assertions[i].result_set[0][0], 900)
+        # wait for processes to return
+        m.wait()
+
+        # get the results
+        result = m.get()
+
+        # validate result.
+        for i in range(CLIENT_COUNT):         
+            self.env.assertEquals(result[i].result_set[0][0], 900)
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:166]: switch to `all`

File: tests/flow/test_concurrent_query.py [hunk 27]
```diff
@@ -172,22 +145,25 @@ def test_05_concurrent_read_delete(self):
...
-            graph = graphs[i]
-            t = threading.Thread(target=thread_run_query, args=(graph, q, i))
-            t.setDaemon(True)
-            threads.append(t)
-            t.start()
+        queries = [q] * CLIENT_COUNT
+        # invoke queries
+        m = pool.amap(thread_run_query, graphs, queries, indexes)
 
         redis_con.delete(GRAPH_ID)
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:156]: This test need to be revisit, we're querying different graphs but are trying to delete only a single graph
I think the idea is to try and delete a graph which is currently being queried multiple times

File: tests/flow/test_concurrent_query.py [hunk 28]
```diff
@@ -246,96 +224,95 @@ def test_05_concurrent_read_delete(self):
...
         redis_con = self.env.getConnection()
         new_graph = GRAPH_ID + "2"
         # Create new empty graph with id GRAPH_ID + "2"
         redis_con.execute_command("GRAPH.QUERY",new_graph, """MATCH (n) return n""", "--compact")
         heavy_write_query = """UNWIND(range(0,999999)) as x CREATE(n)"""
-        writer = threading.Thread(target=thread_run_query, args=(graphs[0], heavy_write_query, 0))
-        writer.setDaemon(True)
-        writer.start()
+        writer = pool.apipe(thread_run_query, graphs[0], heavy_write_query, 0)
         redis_con.rename(new_graph, GRAPH_ID)
```
Comment (swilly22) [inline tests/flow/test_concurrent_query.py:253]: Shouldn't this be: `redis_con.rename(GRAPH_ID, new_graph)`
See: https://redis-py.readthedocs.io/en/stable/index.html?highlight=rename#redis.Redis.rename

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/flow/test_query_mem_limit.py (hunk 11 above)
Comment (AviAvni) [reply tests/flow/test_query_mem_limit.py:66]: yes you see that the for loop build them

File: tests/flow/test_stress.py (hunk 7 above)
Comment (AviAvni) [reply tests/flow/test_stress.py:134]: no

**Review by swilly22 — DISMISSED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**